### ISSUE TITLE
Additional character layouts

### ISSUE THREAD

**Issue by ryansloan** [code-dot-org/dance-party#73]
**In all layouts, Characters will be resized to fit in the view, down to .3 scale**
Sorry for the humongous pictures, I didn't realize when I snapped them and uploaded them that my phone went all out. 

# "Outside Border"
This will create a square border of characters around the screen, with the middle open. It will fill in the following order:
![First Four sprites for outside border](https://user-images.githubusercontent.com/1669271/47760483-b9b71a00-dc71-11e8-8b48-c97034ce1305.png)

After they have more than 8 characters, it will fill in clockwise in the middle of each gap:
![9+ sprites for outside border](https://user-images.githubusercontent.com/1669271/47760503-cfc4da80-dc71-11e8-9c1a-1242f9bf60f6.png)

# "Filled Square"
The filled square layout will grow out from the center. It will fill in starting from the top right, and will proceed right all the way to the last column, then down to the next row. The target square will be MxM where `M = Math.ceil(Math.sqrt(sprites.length))`. Example growth:

![Growth for Filled Square](https://user-images.githubusercontent.com/1669271/47760530-ebc87c00-dc71-11e8-9569-d4e7bad9fa8d.png)

# "Circle" 
The Circle layout already exists, but should be updated to do the following:

* Resize the sprites to fit 
* Rotate them so that they are all angled along the circle's radius at that point. (The circles in the diagram below are the "heads")
* Order such that z-index of sprites grows with their Y position (ex: a sprite that is on the bottom of the canvas at [200, 380] should be on a layered above a sprite that is higher on the canvas at [200,300]
* Place sprites so that they are positioned in steps of 360/N degrees

# "Plus"
The plus layout will grow out from the center, ordering each new character from top->bottom->left->right->top...

![Plus Layout growth](https://user-images.githubusercontent.com/1669271/47761084-5e3a5b80-dc74-11e8-9ce3-9418a64be56b.jpg)

# "X"
The X layout will grow from the center like the plus, but with a few key differences:
* Characters will be positioned 45 degree angles from the origin instead of 90 degree angles
* Characters will be rotated to match their angle of positioning (the circles in the diagram below are the "heads")

![X Layout Growth](https://user-images.githubusercontent.com/1669271/47761317-254eb680-dc75-11e8-9b10-e648b36e60f5.JPG)

# original sketches
The designs above differ slightly from these sketches, because the designs were scaled to allow for an arbitrary number of sprites.

![layouts](https://user-images.githubusercontent.com/1669271/47740430-09c3bb80-dc35-11e8-8dc0-a3d4a230c7f0.png)

**Comment by joshlory** [comment]
Is there any way to generalize this, so the "preset" arrangements are a special case of a more general API?

Reminds me of the https://makecode.microbit.org/ show leds block:

![image](https://user-images.githubusercontent.com/413693/47740743-ca499f00-dc35-11e8-9f74-616b86f31781.png)

Probably not in scope, but want to at least ask the question!

**Comment by ryansloan** [comment]
@joshlory That's an awesome idea! These will need to work with an arbitrary number of sprites though, so I am not sure how well we could define a general API for this. At this point I'd rather work to get the must-have layouts out and then we can try to generalize, if that sounds good?

**Comment by Bjvanminnen** [comment]
Right now the max dancers we allow with this block is 10. Do we expect/desire that to change?

**Comment by Bjvanminnen** [comment]
For circle, arent we going to get better looking circles for many values of N by behaving as we currently doing and just divide 360 by N and have that be the angle between each.

I do like the plan in general around z-ordering.

**Comment by Bjvanminnen** [comment]
Regarding sizing, it's worth pointing out that in the tutorial we resize the dancers to 50 (via another block) after introducing the layout block https://studio.code.org/s/dance/stage/1/puzzle/12. If we're optimizing our shapes for 30, I suspect we'll want to get rid of that resize to 50 block

**Comment by ryansloan** [comment]
> Right now the max dancers we allow with this block is 10. Do we expect/desire that to change?

There are also independent "set layout" blocks that can adjust the layout of an arbitrary number of sprites, so we need it to be generalized



> For circle, arent we going to get better looking circles for many values of N by behaving as we currently doing and just divide 360 by N and have that be the angle between each.

Good call, updated spec.





> Regarding sizing, it's worth pointing out that in the tutorial we resize the dancers to 50 (via another block) after introducing the layout block https://studio.code.org/s/dance/stage/1/puzzle/12. If we're optimizing our shapes for 30, I suspect we'll want to get rid of that resize to 50 block

Yep, we will need to get rid of it. We just have that block there right now because the layout blocks don't do any automatic resizing.

**Comment by joshlory** [comment]
One last question — do we want advanced students to be able to "see inside" the layout blocks?  Again probably out of scope for Hour of Code, but a possible extension for Hour Two.

I remember @mikeharv getting feedback that there isn't a natural use for functions in this tutorial.  Maybe in Hour Two we expose the layout blocks as functions, that act on sprite positions?

**Comment by ryansloan** [comment]
@joshlory That would be cool. Out of scope for Hour of Code, and I wouldn't want to introduce each block separately (like we would want to do if we made them functions) for the HOC tutorial. But I do like the idea of opening the hood on this at some point.

It brings to mind something we have talked about before, enum parameters for student-defined functions. Those would let them create blocks with dropdowns in them. But, I think this is a project for another time.
